src/client/datascience/kernel-launcher/kernelLauncher.ts

```diff
@@ -84,6 +92,42 @@ export class KernelLauncher implements IKernelLauncher {
 
     @captureTelemetry(Telemetry.KernelLauncherPerf)
     public async launch(
+        kernelConnectionMetadata: KernelSpecConnectionMetadata | PythonKernelConnectionMetadata,
+        timeout: number,
+        resource: Resource,
+        workingDirectory: string,
+        cancelToken?: CancellationToken
+    ): Promise<IKernelProcess> {
+        // If this is a python interpreter, make sure it has ipykernel
+        if (kernelConnectionMetadata.interpreter) {
+            await this.installDependenciesIntoInterpreter(kernelConnectionMetadata.interpreter, cancelToken);
+        }
+
+        // Should be available now, wait with a timeout
+        const result = await waitForPromise(
+            Promise.race([
+                this.launchProcess(kernelConnectionMetadata, resource, workingDirectory),
```

Comment:
Shouldn't we also pass the `token` into the `launchProcess`.
If the user cancels, then we should kill the process as well.
Same applies to `timeout`.
If the time span has exceeded, then we should grafefully dispose the kernelProcess that we were attempting to start.

More of a cleanup (since you area cleaning this part).

Reply:
Good idea. I'll investigate. We should dispose the kernelProcess if the cancel happened or the timeout occurred.

Reply:
Again, not necessary, but a good cleanup (cleaning up resources instead of leaving these dangling processes).

Reply:
Should be lower down now. Timeout will happen waiting for the heartbeat. No longer need the waitForPromise stuff.